src/ziterate.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "ZITERATE" "1" "December 2024" "ZMap" "ziterate"
.
.SH "NAME"
\fBziterate\fR \- ZMap IP permutation generation file
.
.SH "SYNOPSIS"
ziterate [ \-b <blocklist> ] [ \-w <allowlist> ] [ OPTIONS\.\.\. ]
.
.SH "DESCRIPTION"
\fIZIterate\fR is a network tool that will produce IPv4 addresses in a psuedorandom order similar to how ZMap generates random addresses to be scanned\.
.
.SH "OPTIONS"
.
.SS "BASIC OPTIONS"
.
.TP
\fB\-p\fR, \fB\-\-target\-ports=port(s)\fR
List of TCP/UDP ports and/or port ranges to scan\. (e\.g\., 80,443,100\-105)\. Use \'*\' to scan all ports, including port 0\. If no port is specified, ziterate will output only IPs\.
.
.TP
\fB\-b\fR, \fB\-\-blocklist\-file=path\fR
File of subnets to exclude, in CIDR notation, one\-per line\. It is recommended you use this to exclude RFC 1918 addresses, multicast, IANA reserved space, and other IANA special\-purpose addresses\. An example blocklist file \fBblocklist\.conf\fR for this purpose\.
.
.TP
\fB\-w\fR, \fB\-\-allowlist\-file=name\fR
File of subnets to include, in CIDR notation, one\-per line\. All other subnets will be excluded\.
.
.TP
\fB\-l\fR, \fB\-\-log\-file=name\fR
File to log to\.
.
.TP
\fB\-\-disable\-syslog\fR
Disable logging messages to syslog\.
.
.TP
\fB\-v\fR, \fB\-\-verbosity\fR
Level of log detail (0\-5, default=3)
.
.TP
\fB\-\-ignore\-blocklist\-errors\fR
Ignore invalid entries in the blocklist\. Default is false\.
.
.TP
\fB\-\-seed=n\fR
Seed used to select address permutation\.
.
.TP
\fB\-n\fR, \fB\-\-max\-targets=n\fR
Cap number of IPs to generate (as a number or a percentage of the address space)
.
.SS "SHARDING"
.
.TP
\fB\-\-shards=n\fR
Total number of shards\.
.
.TP
\fB\-\-shard=n\fR
Shard this scan is targeting\. Zero indexed\.
.
.SS "ADDITIONAL OPTIONS"
.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help text and exit\.
.
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version and exit\.



CHANGELOG
1.0.0 	2013-8-16
Initial public release.

1.0.1	2013-8-17
BUGFIX 	Issue #4 "Missing module_ssldb? Redis module won't compile."
	- removed dependencies on ssldb
SUPPORT	Issue #3 "Error after running make"
	- added documentation that zmap requires 64-bit system
SUPPORT	Issue #1 "Failure at calloc for 'ip_seen' on KVM VPSs?"
	- added documentation on memory requirements for zmap

1.0.2	2013-8-18
BUGFIX	Issue #14 "gcc 4.7.2 -Wunprototyped-calls error with recv_run."
	- changed recv_run header to match def in recv.c

1.0.3	2013-8-19
BUGFIX	Issues 21,28 "fixed get_gateway malloc/memset errors"
BUGFIX  Issue 24 "Makefile does not respect LDFLAGS"
	- changed = to += for CFLAGS, LDFAGS, and LDLIBS


1.1.0	2013-11-18
FEATURE Updated probe and output module interface that allows arbitrary
	data to be passed from the probe module (e.g. additional TCP fields)
	that can then be output as requested.

FEATURE	Replaced simple_file, and redis_file output modules with csv module
	that allows user controlled output of what fields shoudl be output
	to a csv file. As well, implemented --list-output-fields that allows
	users to find what fields are available.	

FEATURE	Added output-filters that allow users to control what types of packets
	that want output (e.g. classification = "SYNACK" && is_repeat = 0)

FEATURE	Drop root privileges after opening necessary sockets if run as
	privileged user.

FEATURE	Added paged bitmap for removing duplicate responses so that if small
	subnets are scanned, large amount of memory is no longer required.

FEATURE	Fast scanning of small subnets. Scanning small subnets no longer
	requires iterating over the entire IPv4 address space, which allows
	ZMap-like speed for all network sizes.

FEATURE Scan CIDR blocks from the command-line instead of only through
	whitelist file (e.g. zmap -p 443 192.168.0.0/16 10.0.0.0/8)

FEATURE Redis output module now allows connecting to arbitrary redis servers
	and lists from the command-line via output module parameter

FEATURE JSON output module added

FEATURE 32-bit support

FEATURE UDP Packet Support

BUGFIX	Source port in UDP module corrected to use network order instead of
	host order

1.1.1	2013-12-16
BUGFIX	Fixed bit-map, which was incorrectly deduplicating responses

BUGFIX	CMake correctly installs files into /etc/zmap

1.1.2	2014-01-21
BUGFIX	Off-by-one error in paged bitmap


1.2.0	2014-03-10
FEATURE added MongoDB as a supported output module

BUGFIX	ICMP values added incorrectly, timestamp unavailable

BUGFIX	setting fixed number of probes may inadverantly target specific networks

BUGFIX	scans occasionally skip cooldown period due to lock issue

FEATURE	added context to allow easier packaging in Debian and Ubuntu and Fedora and Brew and Macports

FEATURE	remove dnet dependency for linux

FEATURE remove random extra printed saddr

FEATURE	build with optimizations by default

FEATURE	added JSON metadata output

FEATURE added syslog support

FEATURE	added BSD/mac support

BUGFIX	removed bizarre executible bits on random example code in git repo

FEATURE	added support for scanning by FQDN

FEATURE	adding sharding support

1.2.1	2014-06-09
FEATURE	autodetect number of sender threads

FEATURE add ignore-invalid-hosts option for blacklist

BUGFIX	UDP scans sometimes double-counted IP header length

BUGFIX	Properly check UDP packet length

BUGFIX	Integer overflow in json metadata when printing generator

BUGFIX	All calls to malloc checked for failure

2.1.0	2015-09-02

BUGFIX  ZMap now filters out packets that are from the local MAC instead of
        only capturing packets from the local gateway. The prior approach
        caused valid responses to be dropped for a fair number of users.

BUGFIX  ZMap would somtimes segfault if the number of threads was greater
        than the number of destination hosts.

BUGFIX  ZMap did not crash when it was unable to write to the output file
        This would cause ZMap to continue running when it was piped into another
        application and that application died. We not log_fatal if the output is
        no longer accessible per ferror.

CHANGED json output uses underscores instead of hyphens

CHANGED removes support for deprecated simple_file and extended_file options

BUGFIX  pcap filter captures outgoing packets

BUGFIX  install overwrites blacklist file

BUGFIX  output is sometimes colored

FEATURE	zblacklist (a standalone utility that allows you to efficiently check
        IP addresses against a zmap compatible whitelist and blacklist. This is
        helpful if you are doing something like cat list-of-ips | zgrab and to
        make sure that you're still following your blacklist.

FEATURE	ztee (a standalone utility that buffers between ZMap and ZGrab) and allows
        extracting just IP address from a larger ZMap output in order to complete
        follow up handshakes without losing any data.

CHANGED	remove space in csv output header

CHANGED	build with JSON support by default

FEATURE	ntp probe module

BUGFIX	finding a generator would crash with low probability

CHANGED	don't print blacklisted CIDR blocks to log. These are available in 
        --metadata-file and end up flooding the log with a ton of metadata.

FEATURE	status-updates-file (monitor output as a csv)

CHANGED	output to stdout by default

BUGFIX	use correct email for Zakir in AUTHORS

FEATURE	add redis-csv output module

CHANGED rename redis module to redis-packed

CHANGED	probe module API takes user data argument

FEATURE	colored log output

FEATURE	add pf_ring and 10GigE support

BUGFIX random-number generator is now thread safe

FEATURE	optional app-success field output in monitor

CHANGED remove type field from json output module and get rid of header

FEATURE Add icmp-echo-time probe module for measuring RTT

CHANGED remove --summary. This has been replaced by --metadata-file.

CHANGED JSON metadata now uses ISO-8601 compatible timestamps instead of proprietary log format.


src/ziterate.c
/*
 * ZMap Copyright 2023 Regents of the University of Michigan
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 */

/*
 * ZIterate is a simple utility that will iteratate over the IPv4
 * space in a pseudo-random fashion, utilizing the sharding capabilities * of
 * ZMap.
 */

#define _GNU_SOURCE

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <getopt.h>
#include <assert.h>
#include <unistd.h>

#include "../lib/includes.h"
#include "../lib/blocklist.h"
#include "../lib/logger.h"
#include "../lib/random.h"
#include "../lib/util.h"
#include "../lib/xalloc.h"

#include "iterator.h"
#include "ports.h"
#include "state.h"
#include "validate.h"
#include "zitopt.h"

struct zit_conf {
	char *blocklist_filename;
	char *allowlist_filename;
	char **destination_cidrs;
	int destination_cidrs_len;
	char *log_filename;
	int check_duplicates;
	int ignore_errors;
	int verbosity;
	int disable_syslog;

	// sharding options
	uint16_t shard_num;
	uint16_t total_shards;
	uint64_t seed;
	aesrand_t *aes;
	uint32_t max_hosts;
};

#define SET_BOOL(DST, ARG)              \
	{                               \
		if (args.ARG##_given) { \
			(DST) = 1;      \
		};                      \
	}

int main(int argc, char **argv)
{
	struct zit_conf conf;

	memset(&conf, 0, sizeof(struct zit_conf));
	conf.verbosity = 3;
	conf.ignore_errors = 0;

	struct gengetopt_args_info args;
	struct cmdline_parser_params *params;
	params = cmdline_parser_params_create();
	assert(params);
	params->initialize = 1;
	params->override = 0;
	params->check_required = 0;

	if (cmdline_parser_ext(argc, argv, &args, params) != 0) {
		exit(EXIT_SUCCESS);
	}

	// Handle help text and version
	if (args.help_given) {
		cmdline_parser_print_help();
		exit(EXIT_SUCCESS);
	}
	if (args.version_given) {
		cmdline_parser_print_version();
		exit(EXIT_SUCCESS);
	}

	// Set the log file and metadata file
	if (args.log_file_given) {
		conf.log_filename = strdup(args.log_file_arg);
	}
	if (args.verbosity_given) {
		conf.verbosity = args.verbosity_arg;
	}
	// Read the boolean flags
	SET_BOOL(conf.ignore_errors, ignore_blocklist_errors);
	SET_BOOL(conf.disable_syslog, disable_syslog);

	// initialize logging
	FILE *logfile = stderr;
	if (conf.log_filename) {
		logfile = fopen(conf.log_filename, "w");
		if (!logfile) {
			fprintf(
			    stderr,
			    "FATAL: unable to open specified logfile (%s)\n",
			    conf.log_filename);
			exit(1);
		}
	}
	if (log_init(logfile, conf.verbosity, !conf.disable_syslog,
		     "ziterate")) {
		fprintf(stderr, "FATAL: unable able to initialize logging\n");
		exit(1);
	}

	// Blocklist and allowlist
	if (args.blocklist_file_given) {
		conf.blocklist_filename = strdup(args.blocklist_file_arg);
	}
	if (args.allowlist_file_given) {
		conf.allowlist_filename = strdup(args.allowlist_file_arg);
	}
	conf.destination_cidrs = args.inputs;
	conf.destination_cidrs_len = args.inputs_num;
	// max targets
	if (args.max_targets_given) {
		conf.max_hosts = parse_max_targets(args.max_targets_arg, zconf.ports->port_count);
	}

	// sanity check blocklist file
	if (conf.blocklist_filename) {
		log_debug("ziterate", "blocklist file at %s to be used",
			  conf.blocklist_filename);
	} else {
		log_debug("ziterate", "no blocklist file specified");
	}
	if (conf.blocklist_filename &&
	    access(conf.blocklist_filename, R_OK) == -1) {
		log_fatal("ziterate",
			  "unable to read specified blocklist file (%s)",
			  conf.blocklist_filename);
	}

	// sanity check allowlist file
	if (conf.allowlist_filename) {
		log_debug("ziterate", "allowlist file at %s to be used",
			  conf.allowlist_filename);
	} else {
		log_debug("ziterate", "no allowlist file specified");
	}
	if (conf.allowlist_filename &&
	    access(conf.allowlist_filename, R_OK) == -1) {
		log_fatal("ziterate",
			  "unable to read specified allowlist file (%s)",
			  conf.allowlist_filename);
	}

	// parse blocklist and allowlist
	if (blocklist_init(conf.allowlist_filename, conf.blocklist_filename,
			   conf.destination_cidrs, conf.destination_cidrs_len,
			   NULL, 0, conf.ignore_errors)) {
		log_fatal("ziterate",
			  "unable to initialize blocklist / allowlist");
	}

	// Set up sharding
	conf.shard_num = 0;
	conf.total_shards = 1;
	if ((args.shard_given || args.shards_given) && !args.seed_given) {
		log_fatal("ziterate",
			  "Need to specify seed if sharding a scan");
	}
	if (args.shard_given ^ args.shards_given) {
		log_fatal(
		    "ziterate",
		    "Need to specify both shard number and total number of shards");
	}
	if (args.shard_given) {
		enforce_range("shard", args.shard_arg, 0, 65534);
		conf.shard_num = args.shard_arg;
	}
	if (args.shards_given) {
		enforce_range("shards", args.shards_arg, 1, 65535);
		conf.total_shards = args.shards_arg;
	}
	if (conf.shard_num >= conf.total_shards) {
		log_fatal("ziterate",
			  "With %hhu total shards, shard number (%hhu)"
			  " must be in range [0, %hhu)",
			  conf.total_shards, conf.shard_num, conf.total_shards);
	}
	log_debug(
	    "ziterate",
	    "Initializing sharding (%d shards, shard number %d, seed %llu)",
	    conf.total_shards, conf.shard_num, conf.seed);

	// Check for a random seed
	if (args.seed_given) {
		conf.seed = args.seed_arg;
	} else {
		if (!random_bytes(&conf.seed, sizeof(uint64_t))) {
			log_fatal("ziterate", "unable to generate random bytes "
					      "needed for seed");
		}
	}
	zconf.aes = aesrand_init_from_seed(conf.seed);

	zconf.ports = xmalloc(sizeof(struct port_conf));
	if (args.target_ports_given) {
		parse_ports(args.target_ports_arg, zconf.ports);
	} else {
		zconf.ports->port_count = 1;
	}

	uint64_t num_addrs = blocklist_count_allowed();
	if (zconf.list_of_ips_filename) {
		log_debug("send",
			  "forcing max group size for compatibility with -I");
		num_addrs = 0xFFFFFFFF;
	}
	iterator_t *it = iterator_init(1, conf.shard_num, conf.total_shards,
				       num_addrs, zconf.ports->port_count);
	shard_t *shard = get_shard(it, 0);

	target_t current = shard_get_cur_target(shard);
	for (uint32_t count = 0; current.ip; ++count) {
		if (conf.max_hosts && count >= conf.max_hosts) {
			break;
		}
		struct in_addr next_ip;
		next_ip.s_addr = current.ip;
		if (current.port) {
			printf("%s,%u\n", inet_ntoa(next_ip), current.port);
		} else {
			printf("%s\n", inet_ntoa(next_ip));
		}
		current = shard_get_next_target(shard);
	}
	return EXIT_SUCCESS;
}
